engine/engine.go

```diff
@@ -738,6 +739,18 @@ func expandStep(s *step.Step, res *resolution.Resolution) {
 	// generate all children steps
 	for i, item := range items {
 		childStepName := fmt.Sprintf("%s-%d", s.Name, i)
+
+		// copy all slices from the parent step to prevent array pointer
```

Comment:
Can we do it only when we follow a `foreach sequence` ?
I don't know of any other situation where we alter the children, and for big expands it could generate a lot of useless garbage.
If you're worried about maintainability, we could have a helper func e.g. `step.ShouldDeepCopy()` ?

Reply:
I'm not that concerned about the garbage that it could creates, I expect all arrays to be relatively small.
Should we proceed with the fix of the bug, and check for alternative optimization later?

Reply:
Agreed with @rbeuque74, I think we can left as-is for the time being, we're not deling with huge arrays.

Reply:
We can't wait anymore, let's prod this